### PR TITLE
[GPT-2] Add docs

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

Lots of people don't seem to know about batched generation with GPT-2 and friends. Hence this PR adds a section to the docs, similar to the T5 docs.

It also fixes an issue in the T5 docs.

### CODE REVIEW

**Comment by sgugger** [comment]
cc @stevhliu

File: docs/source/en/model_doc/gpt2.md [hunk 1]
```diff
@@ -60,6 +60,94 @@ different sizes: small, medium, large, xl and a distilled version of the small c
 
 This model was contributed by [thomwolf](https://huggingface.co/thomwolf). The original code can be found [here](https://openai.com/blog/better-language-models/).
 
+## Training
+
+GPT-2 is a decoder-only Transformer. It is trained using teacher forcing. This means that during training, we feed the ground truth
+previous tokens to the model in order to predict the next one. During training, one needs to provide the input sequences to the model as
+`input_ids`, as well as `labels`. The latter are identical to the `input_ids`. In teacher-forcing style, the model will internally shift the
+logits before calculating the loss. This ensures that the model is trained to predict the next token.
```
Comment (stevhliu) [inline docs/source/en/model_doc/gpt2.md:68]:
```suggestion
GPT-2 is a decoder-only Transformer. It is trained using teacher-forcing which means during training, the previous tokens (the ground truth) are fed to the model to predict the next one. During training, you need to provide the input sequences to the model as
`input_ids` as well as `labels`. The latter is identical to `input_ids` because in teacher-forcing, the model internally shifts the
logits before calculating the loss. This ensures that the model is trained to predict the next token.
```

File: docs/source/en/model_doc/gpt2.md [hunk 2]
```diff
@@ -60,6 +60,94 @@ different sizes: small, medium, large, xl and a distilled version of the small c
...
+
+One can use [`GPT2LMHeadModel`] (or the Tensorflow/Flax variant), which includes the language modeling head on top of the Transformer decoder.
```
Comment (stevhliu) [inline docs/source/en/model_doc/gpt2.md:70]:
```suggestion
You can use [`GPT2LMHeadModel`] (or the TensorFlow/Flax variant), which includes the language modeling head on top of the Transformer decoder.
```

File: docs/source/en/model_doc/gpt2.md [hunk 3]
```diff
@@ -60,6 +60,94 @@ different sizes: small, medium, large, xl and a distilled version of the small c
...
+
+Suppose that we want to train the model for causal language modeling, then examples should be prepared for the model as follows:
```
Comment (stevhliu) [inline docs/source/en/model_doc/gpt2.md:72]:
```suggestion
For example, if you want to train the model for causal language modeling:
```

File: docs/source/en/model_doc/gpt2.md [hunk 4]
```diff
@@ -60,6 +60,94 @@ different sizes: small, medium, large, xl and a distilled version of the small c
...
+>>> input_ids = tokenizer("Hugging Face is a company based in Paris and New York.", return_tensors="pt").input_ids
+>>> labels = input_ids.clone()
+
+>>> # the forward function internally shifts the logits to predict the next tokens and computes a cross-entropy loss
+>>> loss = model(input_ids=input_ids, labels=labels).loss
+>>> loss.item()
+4.075357437133789
+```
+
+As you can see, only 2 inputs are required for the model in order to compute a loss: `input_ids` (which are the
```
Comment (stevhliu) [inline docs/source/en/model_doc/gpt2.md:89]:
```suggestion
As you can see, only 2 inputs are required for the model to compute the loss: `input_ids` (which are the
```

File: docs/source/en/model_doc/gpt2.md [hunk 5]
```diff
@@ -60,6 +60,94 @@ different sizes: small, medium, large, xl and a distilled version of the small c
...
+`input_ids` of the encoded text) and `labels` (which are an exact copy of the `input_ids`, representing the text to be predicted).
+
+However, the example above only shows a single training example. In practice, one trains deep learning models in
+batches. What is typically done is one takes a large corpus of text, concatenates all text into one long sequence, and then chunks them up
+into smaller blocks. Hence GPT-2 style models typically don't require padding or truncation of text during training, as one simply trains on blocks of
+continuous text.
```
Comment (stevhliu) [inline docs/source/en/model_doc/gpt2.md:95]:
```suggestion
However, the example above only shows a single training example. In practice, deep learning models are trained in
batches. Typically, you take a large corpus of text, concatenate all of it into one long sequence, and chunk it 
into smaller blocks. That's why GPT-2 style models typically don't require padding or truncating text during training because it trains on blocks of continuous text.
```

File: docs/source/en/model_doc/gpt2.md [hunk 6]
```diff
@@ -60,6 +60,94 @@ different sizes: small, medium, large, xl and a distilled version of the small c
...
+
+## Inference
+
+At inference time, it is recommended to use [`~generation.GenerationMixin.generate`]. This method takes care of encoding the
+input and auto-regressively generates the next tokens. Check out [this blog post](https://huggingface.co/blog/how-to-generate) to know
+all the details about generating text with Transformers.
```
Comment (stevhliu) [inline docs/source/en/model_doc/gpt2.md:101]: Should we link to this [page](https://huggingface.co/docs/transformers/generation_strategies) instead of the blog? I think it'd be easier to maintain in the docs.

```suggestion
At inference time, you should use [`~generation.GenerationMixin.generate`] to encode the
input and auto-regressively generate the next tokens. Check out [this blog post](https://huggingface.co/blog/how-to-generate) to learn
all the details about generating text.
```

File: docs/source/en/model_doc/gpt2.md [hunk 7]
```diff
@@ -60,6 +60,94 @@ different sizes: small, medium, large, xl and a distilled version of the small c
...
+... Even more surprising to the researchers was the fact that the unicorns spoke perfect English."""
+
+>>> input_ids = tokenizer(prompt, return_tensors="pt").input_ids
+>>> outputs = model.generate(input_ids, max_new_tokens=50)
+>>> print(tokenizer.decode(outputs[0], skip_special_tokens=True))
+In a shocking finding, scientist discovered a herd of unicorns living in a remote, previously unexplored valley, in the Andes Mountains.
+Even more surprising to the researchers was the fact that the unicorns spoke perfect English. "The unicorns were very intelligent, and they were very intelligent," said Dr. David S. Siegel, a professor of anthropology at the University of California, Berkeley. "They were very intelligent, and they were very intelligent, and
+```
+
+The example above only shows a single example. You can also do batched inference, like so:
```
Comment (stevhliu) [inline docs/source/en/model_doc/gpt2.md:119]:
```suggestion
The example above only shows a single example. You can also do batched inference:
```

**Review by stevhliu — APPROVED** [review]
Thanks for writing this up! 👏

I think it may be a better fit for the [Text generation strategies](https://huggingface.co/docs/transformers/generation_strategies) page since you can generalize the batched inference to other GPT2-like models right?

**Comment by NielsRogge** [comment]
@stevhliu can we proceed by adding this, or would you like to have it on a separate page?

**Comment by stevhliu** [comment]
I think it'll be best to put this in its own section on the [Text generation strategies](https://huggingface.co/docs/transformers/generation_strategies#text-generation-strategies) page :)

**Comment by NielsRogge** [comment]
cc @gante will close this PR in favor of adding it to the text generation strategies page.

Would it be possible to add a section on batched generation?

**Comment by gante** [comment]
@NielsRogge absolutely 👍
